### PR TITLE
fix: incorrect proof hash count causing panic

### LINKED ISSUE #20: panic if verifying with incorrect leaf count
When verifying a proof, we rely on the Prover (a potentially malicious party in our case) to inform the Verifier about the correct number of the leaves in the tree. Trying to verify with an incorrect number causes a panic.

Can be reproduced by modifying a line in this example: https://github.com/antouhou/rs-merkle/blob/8ffa623ce70a3659ae73619397b813887cd8d1c9/src/merkle_proof.rs#L155
and changing the last arg to `leaves.len() + 100`

### PR DESCRIPTION
**Pull request by witter-deland**
fix #20

### CODE REVIEW

**Comment by themighty1** [comment]
Hi, would it be possible to also include tests into this PR showcasing what was fixed?

**Comment by witter-deland** [comment]
The problem fixed is a bug, you can refer to #20, I also encountered a similar problem. 

I don't know what kind of test is appropriate to build, it seems to just fix the original bug: use the old code, the merkle proof bug will appear, use the new code, the merkle proof will work normally
> Hi, would it be possible to also include tests into this PR showcasing what was fixed?

**Comment by antouhou** [comment]
@witter-deland Thanks for the work! Yeah, please just include one simple test that passes an incorrect number of hashes, and check that this error is returned. Thank you very much!

**Comment by witter-deland** [comment]
It's done : )
> @witter-deland Thanks for the work! Yeah, please just include one simple test that passes an incorrect number of hashes, and check that this error is returned. Thank you very much!

**Comment by antouhou** [comment]
@witter-deland Thank you! Merged. Going to publish a new patch version soon

**Comment by antouhou** [comment]
Version 1.4.1 published, thank you!

File: tests/merkle_proof_test.rs [hunk 1]
```diff
@@ -80,6 +80,33 @@ pub mod root {
...
+            .iter()
+            .map(|i| leaf_hashes.get(*i).unwrap().clone())
+            .collect();
+
+        let merkle_tree = MerkleTree::<Sha256>::from_leaves(&test_data.leaf_hashes);
+        let proof = merkle_tree.proof(&indices_to_prove);
+        let extracted_root = proof.root(
+            &indices_to_prove,
+            &leaves_to_prove,
+            test_data.leaf_values.len() + 1,
```
Comment (themighty1) [inline tests/merkle_proof_test.rs:101]: @witter-deland, could you pls look into this test again.
the `+1` shouldn't be here iiuc. When I remove it, this test fails.

**Comment by themighty1** [comment]
On second thought, maybe hold off fixing it @witter-deland for now. This PR doesn't seem to address the root cause of the panic. Im working on a PR which addresses it properly. It will subsume this fix.

File: tests/merkle_proof_test.rs (hunk 1 above)
Comment (witter-deland) [reply tests/merkle_proof_test.rs:101]: +1 means incorrect proof hash count, why did you remove it?

Comment (themighty1) [reply tests/merkle_proof_test.rs:101]: i see, sry i was confused for a second. I hope you don't mind if a change the test to actually remove the hash https://github.com/antouhou/rs-merkle/pull/31

**Comment by themighty1** [comment]
nvm, i think this PR's fix covers all cases. thanks.

**Pull request merged**